File: packages/engine/src/lib/cases/declare-king-action.ts
import {
  GameState,
  DeclareKingActionAction,
  Rank,
  logger,
} from '@vinto/shapes';
import copy from 'fast-copy';
import { getTargetTypeFromRank } from '../utils/action-utils';
import {
  clearTossInReadyList,
  getAutomaticallyReadyPlayers,
} from '../utils/toss-in-utils';

/**
 * DECLARE_KING_ACTION Handler
 *
 * Flow:
 * 1. Player has used a King card (via USE_CARD_ACTION)
 * 2. Player has selected a card from their hand or an opponent's hand (via SELECT_ACTION_TARGET)
 * 3. Player declares a rank (any rank including K is allowed)
 * 4. Validate declared rank against the selected card's actual rank
 * 5a. If correct:
 *     - King card goes to discard pile first
 *     - Selected card goes to discard pile
 *     - Trigger toss-in for the declared rank
 * 5b. If incorrect:
 *     - King card goes to discard pile
 *     - Selected card stays visible in hand (marked for animation)
 *     - Player gets a penalty card
 *     - No toss-in is triggered
 * 6. Complete turn (increment turn count, transition to idle)
 *
 * Note: King's ability triggers a toss-in only if the declaration is correct
 */
export function handleDeclareKingAction(
  state: GameState,
  action: DeclareKingActionAction
): GameState {
  const { playerId, declaredRank } = action.payload;

  console.log('[handleDeclareKingAction] Declaring King action:', {
    playerId,
    declaredRank,
  });

  const newState = copy(state);

  // Get selected target
  const selectedTarget = newState.pendingAction?.targets?.[0];
  if (!selectedTarget) {
    logger.error('[handleDeclareKingAction] No target selected');
    return state;
  }

  const { playerId: targetPlayerId, position } = selectedTarget;

  // Find target player and card
  const targetPlayer = newState.players.find((p) => p.id === targetPlayerId);
  if (!targetPlayer) {
    logger.error('[handleDeclareKingAction] Target player not found');
    return state;
  }

  const selectedCard = targetPlayer.cards[position];
  if (!selectedCard) {
    logger.error('[handleDeclareKingAction] Selected card not found');
    return state;
  }

  const actualRank = selectedCard.rank;
  const isCorrect = actualRank === declaredRank;

  console.log('[handleDeclareKingAction] Declaration:', {
    declaredRank,
    actualRank,
    isCorrect,
    targetPlayerId,
    position,
  });

  const isTossInPhase =
    newState.activeTossIn !== null && newState.activeTossIn.queuedActions.length > 0;
  // STEP 1: Move King to discard (ALWAYS happens)
  if (newState.pendingAction?.card) {
    const discardedCard = {
      ...copy(newState.pendingAction.card),
      played: true,
    };
    if (isTossInPhase) {
      // we are in toss-in phase, add card to one below top
      newState.discardPile.addBeforeTop(discardedCard);
    } else {
      newState.discardPile.addToTop(discardedCard);
    }
  }

  // STEP 2: Handle correct vs incorrect
  if (isCorrect) {
    handleCorrectDeclaration(
      newState,
      targetPlayer,
      position,
      selectedCard,
      playerId,
      declaredRank,
      isTossInPhase
    );
  } else {
    handleIncorrectDeclaration(
      newState,
      targetPlayer,
      position,
      selectedCard,
      playerId
    );
  }

  return newState;
}

/**
 * CORRECT Declaration Handler
 *
 * FIX: Properly creates multi-rank toss-in ['K', declaredRank]
 */
function handleCorrectDeclaration(
  newState: GameState,
  targetPlayer: any,
  position: number,
  selectedCard: any,
  playerId: string,
  declaredRank: Rank,
  isTossInPhase: boolean
): void {
  // Remove declared card from hand
  const [removedCard] = targetPlayer.cards.splice(position, 1);

  // Update known positions (shift indices after removal)
  targetPlayer.knownCardPositions = targetPlayer.knownCardPositions
    .filter((pos: number) => pos !== position)
    .map((pos: number) => (pos > position ? pos - 1 : pos));

  // Check if declared card is actionable
  const targetType = getTargetTypeFromRank(selectedCard.rank);

  if (targetType !== undefined) {
    // ACTIONABLE CARD - Setup pending action
    console.log('[handleDeclareKingAction] Correct - actionable card:', {
      rank: removedCard.rank,
      targetType,
    });

    newState.pendingAction = {
      card: removedCard,
      from: 'hand',
      playerId,
      actionPhase: 'choosing-action',
      targetType,
      targets: [],
    };

    // FIX BUG 1 & 3: Properly setup multi-rank toss-in
    setupKingTossInMultiRank(newState, playerId, declaredRank, true);
  } else {
    // NON-ACTIONABLE CARD - Just discard it
    console.log(
      '[handleDeclareKingAction] Correct - non-actionable, discarding'
    );

    if (isTossInPhase) {
      // we are in toss-in phase, add card to one below top
      newState.discardPile.addBeforeTop(removedCard);
    } else {
      newState.discardPile.addToTop(removedCard);
    }
    newState.pendingAction = null;

    // FIX BUG 1 & 3: Properly setup multi-rank toss-in
    setupKingTossInMultiRank(newState, playerId, declaredRank, false);
  }
}

/**
 * INCORRECT Declaration Handler
 */
function handleIncorrectDeclaration(
  newState: GameState,
  targetPlayer: any,
  position: number,
  selectedCard: any,
  playerId: string
): void {
  console.log('[handleDeclareKingAction] INCORRECT declaration');

  const player = newState.players.find((p) => p.id === playerId);
  if (!player) {
    logger.error('[handleDeclareKingAction] Player not found');
    return;
  }

  // Card stays in hand, revealed to ALL players
  for (const p of newState.players) {
    if (p.id === targetPlayer.id) {
      if (!p.knownCardPositions.includes(position)) {
        p.knownCardPositions.push(position);
      }
    } else {
      if (!p.opponentKnowledge) p.opponentKnowledge = {};
      if (!p.opponentKnowledge[targetPlayer.id]) {
        p.opponentKnowledge[targetPlayer.id] = { knownCards: {} };
      }
      p.opponentKnowledge[targetPlayer.id].knownCards[position] = selectedCard;
    }
  }

  // Issue penalty card
  if (newState.drawPile.length > 0) {
    const penaltyCard = newState.drawPile.drawTop();
    if (penaltyCard) {
      player.cards.push(penaltyCard);
      console.log(
        '[handleDeclareKingAction] Penalty card issued:',
        penaltyCard.rank
      );
    }
  }

  newState.pendingAction = null;

  // Setup toss-in for KING ONLY (no declared rank)
  setupKingTossInMultiRank(newState, playerId, undefined, false);
}

/**
 * Setup King multi-rank toss-in
 *
 * FIXES:
 * - BUG 1: Always creates ['K', declaredRank] array (not just ['K'])
 * - BUG 3: APPENDS to existing ranks instead of overwriting
 *
 * @param declaredRank - undefined for incorrect declaration
 * @param hasAction - true if declared card is actionable
 */
function setupKingTossInMultiRank(
  newState: GameState,
  playerId: string,
  declaredRank: Rank | undefined,
  hasAction: boolean
): void {
  // Build rank array: ['K'] or ['K', declaredRank]
  const kingRanks: Rank[] = declaredRank ? ['K', declaredRank] : ['K'];

  if (newState.activeTossIn === null) {
    // START NEW TOSS-IN
    console.log('[setupKingTossInMultiRank] Starting new King toss-in:', {
      ranks: kingRanks,
      hasAction,
    });

    newState.activeTossIn = {
      ranks: kingRanks as [Rank, ...Rank[]],
      initiatorId: playerId,
      originalPlayerIndex: newState.currentPlayerIndex,
      participants: [],
      queuedActions: [],
      waitingForInput: !hasAction,
      playersReadyForNextTurn: getAutomaticallyReadyPlayers(
        newState.players,
        newState.coalitionLeaderId
      ),
    };

    newState.subPhase = hasAction ? 'awaiting_action' : 'toss_queue_active';
  } else {
    // ALREADY IN TOSS-IN - Append King ranks
    console.log(
      '[setupKingTossInMultiRank] Adding King ranks to existing toss-in:',
      {
        existingRanks: newState.activeTossIn.ranks,
        addingRanks: kingRanks,
      }
    );

    // FIX BUG 3: APPEND ranks instead of overwriting
    kingRanks.forEach((rank) => {
      if (!newState.activeTossIn!.ranks.includes(rank)) {
        newState.activeTossIn!.ranks.push(rank);
      }
    });

    // Remove processed King action from queue if applicable
    if (newState.activeTossIn.queuedActions.length > 0) {
      const firstAction = newState.activeTossIn.queuedActions[0];
      if (firstAction.rank === 'K' && firstAction.playerId === playerId) {
        newState.activeTossIn.queuedActions.shift();
        console.log('[setupKingTossInMultiRank] Removed King from queue');
      }
    }

    // Clear ready list (new ranks added)
    clearTossInReadyList(newState);

    newState.subPhase = hasAction ? 'awaiting_action' : 'toss_queue_active';
    newState.activeTossIn.waitingForInput = !hasAction;

    console.log('[setupKingTossInMultiRank] Updated toss-in:', {
      finalRanks: newState.activeTossIn.ranks,
      subPhase: newState.subPhase,
    });
  }
}

File: kmp/shared/engine/src/commonMain/kotlin/game/vinto/engine/cases/DeclareKingAction.kt
package game.vinto.engine.cases

import game.vinto.engine.MutableActiveTossIn
import game.vinto.engine.MutableGameState
import game.vinto.engine.MutablePendingAction
import game.vinto.engine.MutablePlayerState
import game.vinto.engine.PublicReveal
import game.vinto.engine.clearTossInReadyList
import game.vinto.engine.getAutomaticallyReadyPlayers
import game.vinto.engine.getTargetTypeFromRank
import game.vinto.shapes.ActionPhase
import game.vinto.shapes.GameAction
import game.vinto.shapes.GameSubPhase
import game.vinto.shapes.PendingCardOrigin
import game.vinto.shapes.Rank
import game.vinto.shapes.SerializedOpponentKnowledge

/**
 * DECLARE_KING_ACTION — the King names a rank and points at a card.
 *
 * The King is discarded either way. If the declaration was right, the named card leaves the
 * hand and, if it has an action, that action is played; the toss-in then covers *both* King
 * and the declared rank. If it was wrong, the card stays in hand but is revealed to
 * everyone, the player draws a penalty, and only King is tossable.
 *
 * Ported from `packages/engine/src/lib/cases/declare-king-action.ts`.
 */
fun handleDeclareKingAction(state: MutableGameState, action: GameAction.DeclareKingAction): Boolean {
    val playerId = action.payload.playerId
    val declaredRank = action.payload.declaredRank

    val pending = state.pendingAction ?: return false
    val selectedTarget = pending.targets.firstOrNull() ?: return false

    val targetPlayer = state.playerById(selectedTarget.playerId) ?: return false
    val position = selectedTarget.position
    val selectedCard = targetPlayer.cards.getOrNull(position) ?: return false

    val isCorrect = selectedCard.rank == declaredRank

    // Mid-queue, the King must land beneath the top card so an unplayed action stays on top.
    val isTossInPhase = state.activeTossIn?.queuedActions?.isNotEmpty() == true

    val discardedKing = pending.card.copy().also { it.played = true }
    if (isTossInPhase) state.discardPile.addBeforeTop(discardedKing)
    else state.discardPile.addToTop(discardedKing)

    if (isCorrect) {
        applyCorrectDeclaration(state, targetPlayer, position, playerId, declaredRank, isTossInPhase)
    } else {
        applyIncorrectDeclaration(state, targetPlayer, position, playerId)
    }

    return true
}

private fun applyCorrectDeclaration(
    state: MutableGameState,
    targetPlayer: MutablePlayerState,
    position: Int,
    playerId: String,
    declaredRank: Rank,
    isTossInPhase: Boolean,
) {
    val removedCard = targetPlayer.cards.removeAt(position)

    targetPlayer.knownCardPositions.apply {
        val shifted = filter { it != position }.map { if (it > position) it - 1 else it }
        clear()
        addAll(shifted)
    }

    val targetType = getTargetTypeFromRank(removedCard.rank)
    if (targetType != null) {
        state.pendingAction = MutablePendingAction(
            card = removedCard,
            playerId = playerId,
            actionPhase = ActionPhase.CHOOSING_ACTION,
            from = PendingCardOrigin.HAND,
            targetType = targetType,
            targets = mutableListOf(),
        )
        setupKingTossIn(state, playerId, declaredRank, hasAction = true)
    } else {
        if (isTossInPhase) state.discardPile.addBeforeTop(removedCard)
        else state.discardPile.addToTop(removedCard)
        state.pendingAction = null
        setupKingTossIn(state, playerId, declaredRank, hasAction = false)
    }
}

private fun applyIncorrectDeclaration(
    state: MutableGameState,
    targetPlayer: MutablePlayerState,
    position: Int,
    playerId: String,
) {
    val player = state.playerById(playerId) ?: return
    val selectedCard = targetPlayer.cards.getOrNull(position) ?: return

    // Face up for the table, for this moment: naming a card wrongly shows everybody what it
    // really is, and then it goes back to being a card they are expected to remember.
    state.revealed += PublicReveal(targetPlayer.id, position, selectedCard.freeze())

    // The card stays put but everyone now knows it.
    for (observer in state.players) {
        if (observer.id == targetPlayer.id) {
            if (!observer.knownCardPositions.contains(position)) {
                observer.knownCardPositions.add(position)
            }
            continue
        }
        val knowledge = observer.opponentKnowledge ?: mutableMapOf()
        val about = knowledge[targetPlayer.id] ?: SerializedOpponentKnowledge(emptyMap())
        knowledge[targetPlayer.id] =
            about.copy(knownCards = about.knownCards + (position to selectedCard.freeze()))
        observer.opponentKnowledge = knowledge
    }

    if (state.drawPile.length > 0) {
        state.drawPile.drawTop()?.let { player.cards.add(it) }
    }

    state.pendingAction = null
    setupKingTossIn(state, playerId, declaredRank = null, hasAction = false)
}

/**
 * Opens or widens the toss-in for a King.
 *
 * A correct declaration makes both `K` and the declared rank tossable. Crucially the ranks
 * are **appended** to an existing toss-in rather than replacing it: a King played during
 * someone else's toss-in must not cancel the ranks already in play.
 *
 * @param declaredRank null when the declaration was wrong, leaving only King tossable.
 */
private fun setupKingTossIn(
    state: MutableGameState,
    playerId: String,
    declaredRank: Rank?,
    hasAction: Boolean,
) {
    val kingRanks = if (declaredRank != null) listOf(Rank.KING, declaredRank) else listOf(Rank.KING)
    val nextSubPhase =
        if (hasAction) GameSubPhase.AWAITING_ACTION else GameSubPhase.TOSS_QUEUE_ACTIVE

    val tossIn = state.activeTossIn
    if (tossIn == null) {
        state.activeTossIn = MutableActiveTossIn(
            ranks = kingRanks.toMutableList(),
            initiatorId = playerId,
            originalPlayerIndex = state.currentPlayerIndex,
            participants = mutableListOf(),
            queuedActions = mutableListOf(),
            waitingForInput = !hasAction,
            playersReadyForNextTurn = getAutomaticallyReadyPlayers(state),
        )
        state.subPhase = nextSubPhase
        return
    }

    for (rank in kingRanks) {
        if (!tossIn.ranks.contains(rank)) tossIn.ranks.add(rank)
    }

    // This King came off the queue, so take it off.
    tossIn.queuedActions.firstOrNull()?.let { first ->
        if (first.rank == Rank.KING && first.playerId == playerId) {
            tossIn.queuedActions.removeAt(0)
        }
    }

    clearTossInReadyList(state)
    state.subPhase = nextSubPhase
    tossIn.waitingForInput = !hasAction
}
